plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx

```diff
@@ -0,0 +1,288 @@
+		case useSelect.stateChangeTypes.ItemClick:
+			return {
+				...changes,
+				isOpen: true, // Keep menu open after selection.
+				highlightedIndex: state.highlightedIndex,
+			};
+		default:
+			return changes;
+	}
+};
+
+export const CountrySelector = < ItemType extends Item >( {
+	name,
+	className,
+	label,
+	describedBy,
+	options: items,
+	onChange,
+	value,
+	placeholder,
+	children,
+}: ControlProps< ItemType > ): JSX.Element => {
+	const [ searchText, setSearchText ] = useState( '' );
+	const [ isSearchFocused, setSearchFocused ] = useState( false );
+	const [ visibleItems, setVisibleItems ] = useState(
+		new Set( [ ...items.map( ( i ) => i.key ) ] )
+	);
+
+	const itemsToRender = items.filter( ( item ) =>
+		visibleItems.has( item.key )
+	);
+
+	const {
+		getToggleButtonProps,
+		getMenuProps,
+		getItemProps,
+		isOpen,
+		highlightedIndex,
+		selectedItem,
+		closeMenu,
+		selectItem,
+	} = useSelect< ItemType >( {
+		initialSelectedItem: value,
+		items: itemsToRender,
+		stateReducer,
+		onIsOpenChange: () => selectItem( value ),
+	} );
+
+	const itemString = getOptionLabel( value.key, items );
+	const selectedValue = selectedItem ? selectedItem.key : '';
+
+	const searchRef = useRef< HTMLInputElement >( null );
+	const previousStateRef = useRef< {
+		visibleItems: Set< string >;
+	} >();
+
+	function getDescribedBy() {
+		if ( describedBy ) {
+			return describedBy;
+		}
+
+		if ( ! itemString ) {
+			return __( 'No selection', 'woocommerce' );
+		}
+
+		return sprintf(
+			// translators: %s: The selected option.
+			__( 'Currently selected: %s', 'woocommerce' ),
+			itemString
+		);
+	}
+
+	const getSearchSuffix = ( focused: boolean ) => {
+		if ( focused ) {
+			return (
+				<svg
+					xmlns="http://www.w3.org/2000/svg"
+					viewBox="0 0 24 24"
+					width="24"
+					height="24"
+					aria-hidden="true"
+					focusable="false"
+				>
+					<path d="M12 13.06l3.712 3.713 1.061-1.06L13.061 12l3.712-3.712-1.06-1.06L12 10.938 8.288 7.227l-1.061 1.06L10.939 12l-3.712 3.712 1.06 1.061L12 13.061z"></path>
+				</svg>
+			);
+		}
+
+		return (
+			<svg
+				xmlns="http://www.w3.org/2000/svg"
+				viewBox="0 0 24 24"
+				width="24"
+				height="24"
+				aria-hidden="true"
+				focusable="false"
+			>
+				<path d="M13 5c-3.3 0-6 2.7-6 6 0 1.4.5 2.7 1.3 3.7l-3.8 3.8 1.1 1.1 3.8-3.8c1 .8 2.3 1.3 3.7 1.3 3.3 0 6-2.7 6-6S16.3 5 13 5zm0 10.5c-2.5 0-4.5-2-4.5-4.5s2-4.5 4.5-4.5 4.5 2 4.5 4.5-2 4.5-4.5 4.5z"></path>
+			</svg>
+		);
+	};
+
+	// Check if the search input is clearable.
+	const isSearchClearable = isSearchFocused || searchText !== '';
+
+	const menuProps = getMenuProps( {
+		className: 'components-country-select-control__menu',
+		'aria-hidden': ! isOpen,
+	} );
+
+	const onApplyHandler = useCallback(
+		( e: React.MouseEvent< HTMLButtonElement > ) => {
+			e.stopPropagation();
+			onChange( selectedValue );
+			closeMenu();
+		},
+		[ onChange, selectedValue, closeMenu ]
+	);
+
+	const handleSearch = ( {
```

Comment:
I'd debounce this for the fast typists 😄

Reply:
I initially tried to debounce it, but using a ref prop with debounce doesn’t work because the value remains empty due to the delayed state update. Do you have any suggestions to address this?

Reply:
I've pushed some changes in https://github.com/woocommerce/woocommerce/pull/53075/commits/2fe937d01a811d30687f0b95fd7463b1e8ad862d that addresses the debouncing and the i18n accent handling - feel free to adjust or amend

Reply:
Thank you for addressing the debouncing and accents, much appreciated!